src/m3ninx/index/segment/builder/multi_segments_terms_iter.go

```diff
@@ -121,67 +121,68 @@ func (i *termsIterFromSegments) setField(field []byte) error {
+			return false
+		}
 
-	// Create the overlayed postings list for this term
-	i.currPostingsList.Reset()
-	for _, iter := range i.keyIter.CurrentIters() {
-		termsKeyIter := iter.(*termsKeyIter)
-		_, list := termsKeyIter.iter.Current()
-
-		if termsKeyIter.segment.offset == 0 && termsKeyIter.segment.skips == 0 {
-			// No offset, which means is first segment we are combining from
-			// so can just direct union.
-			if err := i.currPostingsList.Union(list); err != nil {
-				i.err = err
-				return false
-			}
-			continue
+		if !i.keyIter.Next() {
+			return false
 		}
 
-		// We have to take into account offset and duplicates/skips.
-		var (
-			iter            = list.Iterator()
-			negativeOffsets = termsKeyIter.segment.negativeOffsets
-			multiErr        = xerrors.NewMultiError()
-		)
-		for iter.Next() {
-			curr := iter.Current()
-			negativeOffset := negativeOffsets[curr]
-			// Then skip the individual if matches.
-			if negativeOffset == -1 {
-				// Skip this value, as itself is a duplicate.
+		// Create the overlayed postings list for this term
+		i.currPostingsList.Reset()
+		for _, iter := range i.keyIter.CurrentIters() {
+			termsKeyIter := iter.(*termsKeyIter)
+			_, list := termsKeyIter.iter.Current()
+
+			if termsKeyIter.segment.offset == 0 && termsKeyIter.segment.skips == 0 {
+				// No offset, which means is first segment we are combining from
+				// so can just direct union.
+				if err := i.currPostingsList.Union(list); err != nil {
+					i.err = err
+					return false
+				}
 				continue
 			}
-			value := curr + termsKeyIter.segment.offset - postings.ID(negativeOffset)
-			if err := i.currPostingsList.Insert(value); err != nil {
-				multiErr = multiErr.Add(err)
-				multiErr = multiErr.Add(iter.Close())
-				i.err = multiErr.FinalError()
+
+			// We have to take into account offset and duplicates/skips.
+			var (
+				iter            = list.Iterator()
+				negativeOffsets = termsKeyIter.segment.negativeOffsets
+				multiErr        = xerrors.NewMultiError()
+			)
+			for iter.Next() {
+				curr := iter.Current()
+				negativeOffset := negativeOffsets[curr]
+				// Then skip the individual if matches.
+				if negativeOffset == -1 {
+					// Skip this value, as itself is a duplicate.
+					continue
+				}
+				value := curr + termsKeyIter.segment.offset - postings.ID(negativeOffset)
+				if err := i.currPostingsList.Insert(value); err != nil {
+					multiErr = multiErr.Add(err)
+					multiErr = multiErr.Add(iter.Close())
+					i.err = multiErr.FinalError()
+					return false
+				}
+			}
+
+			multiErr = multiErr.Add(iter.Err())
+			multiErr = multiErr.Add(iter.Close())
+			i.err = multiErr.FinalError()
+			if i.err != nil {
 				return false
 			}
 		}
 
-		multiErr = multiErr.Add(iter.Err())
-		multiErr = multiErr.Add(iter.Close())
-		i.err = multiErr.FinalError()
-		if i.err != nil {
-			return false
+		// Continue looping only if everything skipped or term is empty.
+		if !i.currPostingsList.IsEmpty() {
+			break
```

Comment:
nit: Make this a `return true` perhaps?

Then Go will actually let you remove the `return true` from the end of the function since it knows.

Reply:
👍